quickwit-metastore/src/lib.rs

```diff
@@ -59,3 +63,68 @@ pub use metastore::{
     IndexMetadata, MetadataSet, Metastore, SplitMetadata, SplitMetadataAndFooterOffsets, SplitState,
 };
 pub use metastore_resolver::{MetastoreFactory, MetastoreUriResolver};
+
+/// A trait to allow for our data model to check itself
+#[async_trait]
+pub trait Check {
+    /// An async function that implements the check on this type
+    async fn check(&self) -> Result<(), anyhow::Error>;
+}
+
+#[cfg(any(test, feature = "testsuite"))]
+#[async_trait]
+impl Check for MockMetastore {
+    async fn check(&self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+/// Decorate a result as a check (for use in [`do_checks`]).
+///
+/// This takes out the original error of the `Result` (if any), replacing it with "taken".
+/// This measure is unfortunately required by the fact that `anyhow::Error` is not `Clone`.
+/// As the checks should bail anyway, this should not be a problem.
+pub fn checked<T>(r: &mut anyhow::Result<T>) -> Pin<Box<(dyn Future<Output = anyhow::Result<()>> + Send)>> {
```

Comment:
A lot of the extra complexity we are dealing with here is coming from the fact that your check function takes Futures.
Maybe it could just take resolved `anyhow::Result<()>` directly?

The checked function would become irrelevant.
Granted: the CLI display experience will not exactly be as nice.

What do you think @fmassot  @llogiq ?

Reply:
The idea behind `do_checks` is that we can run multiple futures concurrently, which will return results faster when e.g. waiting for a database or S3 instance. So we have this one (admittedly not so small) piece of complexity, but the API is workable (though I might add some more docs) and makes calling it fairly painless.

Reply:
I like the async checks for the reason mentioned by @llogiq.
But... I found that the complexity is a consequence of mixing connectivity checks with parsing checks. Here the source config can be bad only if the source file is badly written. We have the same thing for index config.

What I would do here is checking only connectivity stuff in `do_checks` and leave the parsing checking as it is today. (we could also do a parsing report but I think the issue was not opened for that reason).

```
let source_config =
        create_source_config_from_args(source_config_path_opt, input_path_opt).await?;
do_checks([
            ("source", source_check),
            ("metastore", metastore.check()),
            ("index", metastore.check_index(&args.index_id)),
        ])
        .await?;
```

Reply:
Yes, that makes sense. This would let us remove the `checked` method as well as some rather complex code in `quickwit-cli/src/lib.rs`.